optuna/samplers/_nsga2.py

```diff
@@ -253,51 +288,82 @@ def _select_parent(self, study: Study, population: List[FrozenTrial]) -> FrozenT
-            dominates_list[q.number].append(p.number)
-            dominated_count[p.number] += 1
-
-    population_per_rank = []
-    while population:
-        non_dominated_population = []
-        i = 0
-        while i < len(population):
-            if dominated_count[population[i].number] == 0:
-                individual = population[i]
-                if i == len(population) - 1:
-                    population.pop()
+    def _fast_non_dominated_sort(
+        self,
+        population: List[FrozenTrial],
+        directions: List[optuna.study.StudyDirection],
+    ) -> List[List[FrozenTrial]]:
+        dominated_count: DefaultDict[int, int] = defaultdict(int)
+        dominates_list = defaultdict(list)
+
+        dominates = _dominates if self._constraints_func is None else _constrained_dominates
+
+        for p, q in itertools.combinations(population, 2):
+            if dominates(p, q, directions):
+                dominates_list[p.number].append(q.number)
+                dominated_count[q.number] += 1
+            elif dominates(q, p, directions):
+                dominates_list[q.number].append(p.number)
+                dominated_count[p.number] += 1
+
+        population_per_rank = []
+        while population:
+            non_dominated_population = []
+            i = 0
+            while i < len(population):
+                if dominated_count[population[i].number] == 0:
+                    individual = population[i]
+                    if i == len(population) - 1:
+                        population.pop()
+                    else:
+                        population[i] = population.pop()
+                    non_dominated_population.append(individual)
                 else:
-                    population[i] = population.pop()
-                non_dominated_population.append(individual)
-            else:
-                i += 1
+                    i += 1
 
-        for x in non_dominated_population:
-            for y in dominates_list[x.number]:
-                dominated_count[y] -= 1
+            for x in non_dominated_population:
+                for y in dominates_list[x.number]:
+                    dominated_count[y] -= 1
 
-        assert non_dominated_population
-        population_per_rank.append(non_dominated_population)
+            assert non_dominated_population
+            population_per_rank.append(non_dominated_population)
 
-    return population_per_rank
+        return population_per_rank
+
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        if self._constraints_func is not None:
+            constraints = None
+            try:
+                con = self._constraints_func(trial)
+                if not isinstance(con, (tuple, list)):
+                    warnings.warn(
+                        f"Constraints should be a sequence of floats but got {constraints}."
```

Comment:
Sorry this logic probably comes from the `BoTorchSampler` which I wrote but the warning message is somewhat misleading. Could be a follow up but I just created https://github.com/optuna/optuna/pull/2197 for reference

Reply:
Ah, `constraints` will be `None` here, and we should show the type of `con`. I applied your suggestion in the commit https://github.com/optuna/optuna/pull/2175/commits/18a0c9e9074b19d7f1d02b71a3031502b4420037.

Reply:
Thanks for your quick move!